src/app/services/csvs.service.ts

```diff
@@ -0,0 +1,103 @@
+import { Injectable } from '@angular/core';
+import * as CSVParse from 'papaparse';
+import ZipFile from 'jszip';
+import sanitizeFileName from 'sanitize-filename';
+import { formatISO } from 'date-fns';
+
+type Fixes = { prefix: string; suffix: string };
+
+@Injectable({
+    providedIn: 'root'
+})
+export class CSVsService {
+    private zipObject = new ZipFile();
+    private currentDate: Date | undefined;
+
+    private baselineParseConfig = {
+        skipEmptyLines: true,
+        dynamicTyping: false
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    constructor() {}
+
+    // TODO: - Parsing from & Unparsing to Excel formatted CSVs (https://www.papaparse.com/faq#encoding)
+    //       - Expose Parser Errors & Meta info (`results.errors` & `results.meta`)
+    //       - Update parser to use a stream to parse row by row (makes progress indeterminate)
+
+    async parseCSV(file: File, options?: object) {
+        let promiseResolve: (results: CSVParse.ParseResult<unknown>) => void;
+        const promise = new Promise<CSVParse.ParseResult<unknown>>((resolve) => {
+            promiseResolve = resolve;
+        });
+        CSVParse.parse(file, {
+            ...this.baselineParseConfig,
+            ...options,
+            complete: async (results) => {
+                promiseResolve(results);
+            }
+        });
+        return promise;
+    }
+
+    public async makeFile(
+        namesAndContent: Map<string, Record<string, string>[]>,
+        fixes?: Fixes,
+        zipName?: string,
+        zipFixes?: Fixes
+    ): Promise<File> {
+        this.currentDate = undefined;
+        if (namesAndContent.size === 0) {
+            throw new Error('No filename or data provided to make a file');
+        } else if (namesAndContent.size === 1 && zipName === undefined && zipFixes === undefined) {
+            return this.makeCSVFile(namesAndContent.entries().next().value, fixes);
+        } else {
+            return await this.makeZipFile(namesAndContent, fixes, zipName, zipFixes);
+        }
+    }
+
+    private makeCSVFile(nameAndContent: [string, Record<string, string>[]], fixes?: Fixes) {
+        const [filename, data] = nameAndContent;
+        return new File([CSVParse.unparse(data)], sanitizeFileName(this.filenameTemplate(filename, fixes) + '.csv'), {
+            type: 'text/csv;charset=utf-8;'
+        });
+    }
+
+    private async makeZipFile(
+        namesAndContent: Map<string, Record<string, string>[]>,
+        fixes?: Fixes,
+        zipName?: string,
+        zipFixes?: Fixes
+    ) {
+        const zipFile = new ZipFile();
+        this.zipObject = zipFile;
+        for (const [filename, data] of namesAndContent.entries()) {
+            zipFile.file(sanitizeFileName(this.filenameTemplate(filename, fixes) + '.csv'), CSVParse.unparse(data));
```

Comment:
Is it necessary to applying the prefix & suffix as well as the datetime string to each individual CSV file? Since they are context-dependent (due to the existence of `zipName` and `zipeFixes`), maybe we could just use the provided filename.

Reply:
Name overlapping is something I've never considered. It makes sense, and thanks for pointing it out!